core/CronArchive.php

```diff
@@ -561,663 +418,312 @@ public function run()
-            return;
-        }
-
-        $this->logSection("SUMMARY OF ERRORS");
-        foreach ($this->errors as $error) {
-            // do not logError since errors are already in stderr
-            $this->logger->info("Error: " . $error);
+        foreach ($archiveArray as $entry) {
+            if ($entry['idsite'] == $archive['idsite']
+                && $entry['period'] == $archive['period']
+                && $entry['date1'] == $archive['date1']
+                && $entry['date2'] == $archive['date2']
+                && $entry['name'] == $archive['name']
+            ) {
+                return true;
+            }
         }
-
-        $summary = count($this->errors) . " total errors during this script execution, please investigate and try and fix these errors.";
-        $this->logFatalError($summary);
+        return false;
     }
 
-    public function logFatalError($m)
+    private function getNextInvalidatedArchive($periodToGet)
     {
-        $this->logError($m);
+        $tables = $this->getTablesWithInvalidatedArchives();
 
-        throw new Exception($m);
+        foreach ($tables as $table) {
+            $tableMonth = substr($table, strlen($table) - 7, 7);
+            $tableMonth = str_replace('_', '-', $tableMonth);
+
+            if (!isset($this->idArchivesToExclude[$tableMonth])) {
+                $this->idArchivesToExclude[$tableMonth] = [];
+            }
+
+            $nextArchive = $this->model->getNextInvalidatedArchive($table, $periodToGet, $this->allWebsites, $this->idArchivesToExclude[$tableMonth]);
+            if (!empty($nextArchive)) {
+                $this->findSegmentForArchive($nextArchive);
+
+                return $nextArchive;
+            }
+
+            $this->removeTableThatHasNoInvalidatedArchives($table);
+            unset($this->idArchivesToExclude[$tableMonth]);
+        }
+
+        return null;
     }
 
-    /**
-     * @param int[] $idSegments
-     */
-    public function setSegmentsToForceFromSegmentIds($idSegments)
+    private function getTablesWithInvalidatedArchives()
     {
-        /** @var SegmentEditorModel $segmentEditorModel */
-        $segmentEditorModel = StaticContainer::get('Piwik\Plugins\SegmentEditor\Model');
-        $segments = $segmentEditorModel->getAllSegmentsAndIgnoreVisibility();
+        $cacheKey = self::TABLES_WITH_INVALIDATED_ARCHIVES;
 
-        $segments = array_filter($segments, function ($segment) use ($idSegments) {
-            return in_array($segment['idsegment'], $idSegments);
-        });
+        /** @var Lazy $cache */
+        $cache = Cache::getLazyCache();
+        $result = $cache->fetch($cacheKey);
+        $result = @json_decode($result, true);
 
-        $segments = array_map(function ($segment) {
-            return $segment['definition'];
-        }, $segments);
+        if (!is_array($result)) {
+            $this->invalidateArchivedReportsForSitesThatNeedToBeArchivedAgain();
```

Comment:
(answered below because github had some issues, but now they're gone so answering here just in case it was missed)

It should, yes. Though in this case the invalidation wouldn't have an effect, not sure if we want it to be scheduled for re-archiving again? Probably not... We could also not invalidate archives set to DONE_IN_PROGRESS.

Reply:
> not sure if we want it to be scheduled for re-archiving again?

I reckon we wouldn't want to schedule it again.

> We could also not invalidate archives set to DONE_IN_PROGRESS.

That could work too we just need to keep remembering the flag for the invalidation